DataDog/dd-trace-java
# integrate APM Test Agent within CircleCI instrumentation tests

**What does this PR do?**
This PR adds the APM Test Agent to run alongside integration/plugin tests within CI testing, handling all produced traces during a test run and asserting with any enabled checks. This change was made by modifying the AgentTestRunner to have to contain an additional DDAgentWriter instance, only if the `CI_USE_TEST_AGENT` env variable is set. After test assertions are complete, traces are written from the original ListWriter used for testing to the DDAgentWriter, then flushed to the Test-Agent. 

NOTE: Traces are written and flushed to the DDAgentWriter after all test assertions, as opposed to using the MultiWriter class as it would result in trace / span processing, causing instrumentation tests to fail.

 The APM Test Agent is being added to every DD APM Tracer with the primary purpose of providing a centralized location for writing tests which will subsequently be used for asserting on handled APM traces, APM stats, and/or APM Telemetry. This strategy will help bring consistency to Datadog APM Tagging/ integration support.  [More information on the APM Test Agent is available here.](https://github.com/DataDog/dd-apm-test-agent)

**Motivation**
This testing strategy will be used by Service Naming, Unified Naming Conventions tagging, etc as a centralized location for asserting on integration traces. It allows us to leverage traces produced from integrations during testing, and assert consistently across APM tracers.

**How to test the change?**
Verify via logs that the Test-Agent is receiving traces from instrumentation tests [here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/25235/workflows/62d91df2-7241-4366-be47-bc52093fbf82/jobs/778542?invite=true#step-2-974).
Verify via logs that the Test-Agent is asserting on traces with Trace Checks [here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/25235/workflows/62d91df2-7241-4366-be47-bc52093fbf82/jobs/778542?invite=true#step-118-26).
